### PR TITLE
Sort: Various fixes and performance improvements

### PR DESCRIPTION
**Pull request by electricboogie**
Lots of cleanup.  Fixes to trailing decimals (4568...9099 should be 4568.9099 and not a string for compare), fixes default sort dedup, and COW for removal of decimals and thousands separators, which makes us fast again! Shouldn't conflict with @miDeb too much.  Thanks.

### CODE REVIEW

**Comment by sylvestre** [comment]
which arg did you use to evaluate the perf wins? Thanks

File: src/uu/sort/src/sort.rs [hunk 1]
```diff
@@ -603,12 +606,13 @@ fn default_compare(a: &str, b: &str) -> Ordering {
 #[inline(always)]
 fn leading_num_common(a: &str) -> &str {
     let mut s = "";
+
+    // check whether char is numeric, whitespace or decimal point or thousand seperator
```
Comment (miDeb) [inline src/uu/sort/src/sort.rs:610]:
```suggestion
    // check whether char is numeric, whitespace or decimal point or thousand separator
```
small typo :)

File: src/uu/sort/src/sort.rs [hunk 2]
```diff
@@ -731,20 +738,50 @@ fn get_nums_dedup(a: &str) -> &str {
         ""
     // Prepare lines for comparison of only the numerical leading numbers
     } else {
-        get_leading_num(s)
+        let result = match settings.mode {
+            SortMode::Numeric => get_leading_num(s),
+            SortMode::GeneralNumeric => get_leading_gen(s),
+            SortMode::HumanNumeric => get_leading_num(s),
+            SortMode::Version => get_leading_num(s),
```
Comment (miDeb) [inline src/uu/sort/src/sort.rs:745]: You could use the `|` syntax for patterns here (https://doc.rust-lang.org/nightly/book/ch18-03-pattern-syntax.html#multiple-patterns) and do `SortMode::Numeric | SortMode::HumanNumeric | SortMode::Version => get_leading_num(s)`.

Comment (electricboogie) [reply src/uu/sort/src/sort.rs:745]: Thanks!

**Comment by electricboogie** [comment]
> which arg did you use to evaluate the perf wins? Thanks

I used our numeric compare (-n) which is actually a tad slower than our general numeric (-g) but which emulates GNU/BSD (as far as I know now) relatively correctly.  When I fixed thousand separators to be removed only for numeric compare (to emulate GNU/BSD), removal/replacement of thousand separators within get_leading_num seemed to require we return a string, which we now can avoid.  

This blog entry was a great help: https://lise-henry.github.io/articles/optimising_strings.html

The real win compared to GNU/BSD is not being CPU as much - [User: 2.050 s, System: 0.054 s] vs. [User: 366.9 ms, System: 43.4 ms]

But wall clock speed up is nice too:
 
Summary
  'cat ../countwords/kjvbible-random.txt  | ./target/release/sort -n' ran
    6.62 ± 0.41 times faster than 'cat ../countwords/kjvbible-random.txt  | gsort -n'
    
Current repo (but with naive trailing decimal and thousands only for numeric fixes): 

Summary
  'cat ../countwords/kjvbible-random.txt  | ./target/release/sort -n' ran
    2.64 ± 0.07 times faster than 'cat ../countwords/kjvbible-random.txt  | gsort -n'

**Comment by sylvestre** [comment]
OK, thanks. I tried with hyperfine and it is faster indeed :)
```
$ hyperfine --warmup 2   --show-output 'cat sample.txt|/tmp/coreutils-old sort -n  > /dev/null' 'cat sample.txt|sort -n  > /dev/null' 'cat sample.txt|~/dev/debian/coreutils/target/release/coreutils sort -n   > /dev/null'
Benchmark #1: cat sample.txt|/tmp/coreutils-old sort -n  > /dev/null
  Time (mean ± σ):     756.5 ms ±  45.6 ms    [User: 693.7 ms, System: 75.3 ms]
  Range (min … max):   723.9 ms … 849.9 ms    10 runs
 
Benchmark #2: cat sample.txt|sort -n  > /dev/null
  Time (mean ± σ):      1.369 s ±  0.040 s    [User: 1.329 s, System: 0.045 s]
  Range (min … max):    1.315 s …  1.446 s    10 runs
 
Benchmark #3: cat sample.txt|~/dev/debian/coreutils/target/release/coreutils sort -n   > /dev/null
  Time (mean ± σ):     609.9 ms ±  23.6 ms    [User: 2.621 s, System: 0.089 s]
  Range (min … max):   585.3 ms … 670.2 ms    10 runs
 
Summary
  'cat sample.txt|~/dev/debian/coreutils/target/release/coreutils sort -n   > /dev/null' ran
    1.24 ± 0.09 times faster than 'cat sample.txt|/tmp/coreutils-old sort -n  > /dev/null'
    2.24 ± 0.11 times faster than 'cat sample.txt|sort -n  > /dev/null'
```

**Comment by jhscheer** [comment]
Hi, this PR inadvertently added '.DS_Store' files to the repo.

**Comment by sylvestre** [comment]
removed, thanks

**Pull request merged**